homeassistant/components/decora_wifi/light.py

```diff
@@ -20,85 +16,76 @@
+from homeassistant.helpers.device_registry import DeviceInfo
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
 
+from . import DecoraWifiAsyncClient
+from .const import DOMAIN
+
 _LOGGER = logging.getLogger(__name__)
 
 # Validation of the user's configuration
 PLATFORM_SCHEMA = LIGHT_PLATFORM_SCHEMA.extend(
     {vol.Required(CONF_USERNAME): cv.string, vol.Required(CONF_PASSWORD): cv.string}
 )
 
-NOTIFICATION_ID = "leviton_notification"
-NOTIFICATION_TITLE = "myLeviton Decora Setup"
-
 
-def setup_platform(
+async def async_setup_platform(
     hass: HomeAssistant,
     config: ConfigType,
     add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
-    """Set up the Decora WiFi platform."""
+    """Set up decora_wifi component."""
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
+        )
+    )
 
-    email = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    session = DecoraWiFiSession()
 
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Decora WiFi platform."""
+
+    session: DecoraWiFiSession = hass.data[DOMAIN][entry.entry_id]
+    async_session = DecoraWifiAsyncClient(session, hass)
     try:
-        success = session.login(email, password)
-
-        # If login failed, notify user.
-        if success is None:
-            msg = "Failed to log into myLeviton Services. Check credentials."
-            _LOGGER.error(msg)
-            persistent_notification.create(
-                hass, msg, title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID
-            )
-            return
-
-        # Gather all the available devices...
-        perms = session.user.get_residential_permissions()
-        all_switches: list = []
-        for permission in perms:
-            if permission.residentialAccountId is not None:
-                acct = ResidentialAccount(session, permission.residentialAccountId)
-                all_switches.extend(
-                    switch
-                    for residence in acct.get_residences()
-                    for switch in residence.get_iot_switches()
-                )
-            elif permission.residenceId is not None:
-                residence = Residence(session, permission.residenceId)
-                all_switches.extend(residence.get_iot_switches())
-
-        add_entities(DecoraWifiLight(sw) for sw in all_switches)
-    except ValueError:
-        _LOGGER.error("Failed to communicate with myLeviton Service")
+        permissions = await async_session.get_permissions()
+        residences = await async_session.get_residences(permissions)
+        iot_switches = await async_session.get_iot_switches(residences)
```

Comment:
Ideally we do these calls in `__init__.py` and make `light.py` only responsible for adding the lights from the data we acquired in `init.py`